### PR TITLE
Add Buildkit Option

### LINKED ISSUE #422: Add BuildKit support for Docker Image
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

The Docker `Image` Resource on `latest` is implemented using the Docker client. To use BuildKit, we have to set a mode on the Docker client.
My suggestion after talking to several stakeholders is that we implement BuildKit support by default, giving our users the option to turn this off via passing an env var. This would basically be the [opposite of the example given in the 3.x.x docs](https://www.pulumi.com/registry/packages/docker/api-docs/image/#env_nodejs).

Steps needed:
- Run the Docker client in BuildKit mode or heritage mode using `BuilderVersion`, depending on j
- ~Add input parsing to use `env`~ Create new field named `builderVersion` to determine which mode of the client we're running
- Add tests
- Add documentation

### PR DESCRIPTION
**Pull request by guineveresaenger**
This PR makes an opinionated choice about Docker Buildkit.
Background: The Docker client can run in two separate modes, `BuilderV1` which is the classic Docker builder, and `BuilderBuildKit`, which is the version of the client that providers BuildKit support.
These modes are set via the `Version` field on the ImageBuildOptions. When this field is unset or nil, the Docker client runs in `BuilderV1` mode by default.

In general, it appears as though folks are moving away from using the older Docker builder, and therefore this PR implements the buildkit mode as the default behavior for this resource.
Users may still set their `builderVersion` to the `BuilderV1` mode, but must do so explicitly.
Invalid builder versions will return an error on `pulumi up`.

This functionality in v3.x.x was handled via the `env` field, which asked users to provide `DOCKER_BUILDKIT=1` as an environment variable for the Docker CLI to pick up.
Since this version of `Image` relies on the Docker client, not the CLI, this particular field is no longer necessary in this form.
I believe making `builderVersion` its own schematized type field is a lot more user friendly, more tightly scoped, and allows us to explicitly document this feature.
Furthermore, we will be able to deprecate the `env` field safely, and it is additionally the hope of this PR author that `builderVersion` is much more intuitive for users than `env`

Fixes #422.

- add buildkit types field
- Add logic for Builder Version on client. Add input marshaling tests
- Update Image description
- Regenerate schema and SDKs

### CODE REVIEW

**Comment by viveklak** [comment]
> This functionality in v3.x.x was handled via the env field, which asked users to provide DOCKER_BUILDKIT=1 as an environment variable for the Docker CLI to pick up.

I imagine there might be other uses of the `env` field to pass environment variables/flags to the builder? Or does invoking through the docker client not really recognize those environment variables anyway? That might help define whether we deprecate the field. I agree with your reasoning that specifying the builder explicitly is more intuitive.

File: provider/cmd/pulumi-resource-docker/schema.json [hunk 1]
```diff
@@ -3039,6 +3039,11 @@
                     },
                     "description": "An optional map of named build-time argument variables to set during the Docker build. This flag allows you to pass built-time variablesthat can be accessed like environment variables inside the RUN instruction."
                 },
+                "builderVersion": {
+                    "type": "string",
```
Comment (viveklak) [inline provider/cmd/pulumi-resource-docker/schema.json:3043]: Should we make this an enum instead? https://www.pulumi.com/docs/guides/pulumi-packages/schema/#complextype

File: provider/resources.go [hunk 2]
```diff
@@ -199,6 +199,15 @@ func Provider() tfbridge.ProviderInfo {
 							Description: "The target of the Dockerfile to build",
 							TypeSpec:    schema.TypeSpec{Type: "string"},
 						},
+						"builderVersion": {
+							Description: "The version of the Docker builder. " +
+								"Valid inputs are: \n" +
+								"`BuilderV1` - the first generation builder in docker daemon\n" + "" +
+								"`BuilderBuildKit - the builder based on moby/buildkit project\n " +
+								"Defaults to `BuilderBuildKit`.",
+							TypeSpec: schema.TypeSpec{Type: "string"},
```
Comment (viveklak) [inline provider/resources.go:208]: See above comment on perhaps representing this as an enum instead?

**Comment by t0yv0** [comment]
I have filed a feature request for pulumi/pulumi/sdk/go to build some helper methods to simplify or eliminate the marshaling code here https://github.com/pulumi/pulumi/issues/11530

File: provider/image_test.go [hunk 3]
```diff
@@ -162,7 +163,7 @@ func TestMarshalBuildAndApplyDefaults(t *testing.T) {
 			"extraOptions": resource.NewArrayProperty([]resource.PropertyValue{}),
 		})
 
-		actual := marshalBuildAndApplyDefaults(input)
+		actual, _ := marshalBuildAndApplyDefaults(input)
```
Comment (t0yv0) [inline provider/image_test.go:166]: A bit more verbose but could use

```
actual, err := ...
assert.NoError(t, err)
```

**Comment by t0yv0** [comment]
This makes a lot of sense and LGTM except I echo Vivek's desire to try to make it an enum for better discoverability. I also wonder if we accumulate this in pending release notes somewhere? This would be great to highlight to users as part of release notes ("here's how it used to work, and here is how it works now").

**Comment by guineveresaenger** [comment]
@t0yv0 - yes! pretty much [exactly this](https://pkg.go.dev/github.com/docker/docker/api/types@v20.10.21+incompatible#ImageBuildOptions.BuildArgs) for `ImageBuildOptions` which [we pass to the Docker client](https://github.com/pulumi/pulumi-docker/blob/master/provider/image.go#L93).

**Pull request merged**